### PR TITLE
Create CLI Tool. Improve Loading Bar on visual

### PR DESCRIPTION
**Pull request by ivanseidel**
Two changes were made:

1. Bar of method 'visual' now has a nicer look and works better (See: https://asciinema.org/a/2cv438dcpbj3fxj4x0yhmbg8l)
2. A CLI tool was added. Now it's possible to install it globally, and execute `speedtest-net` to do test. Here is the CLI UI:

CLI Tool

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,17 @@
-speedtest.net client
+# speedtest.net API / CLI tool
+
+![SpeedTest Cli](midia/speedtestcli.gif)
```
Comment (ddsol) [inline README.md]: Could you rename `midia` to `media`?

File: bin/index.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env node
+var DraftLog = require('draftlog').into(console).addLineListener(process.stdin)
```
Comment (ddsol) [inline bin/index.js]: Would you mind you adding `;` to every line and make each `if` have a block `{ }` to keep consistent code style?

File: bin/index.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+var SpeedTestNet = require('../')
+
+const DIM = '\033[2m'
```
Comment (ddsol) [inline bin/index.js]: Might it be better to use [`chalk`](https://github.com/chalk/chalk) for consistent color support?

Comment (ivanseidel) [inline bin/index.js]: I avoid adding dependencies if they are not totally necessary. If you want, I can change to `chalk`, but would be cool to keep deps as low as possible (it's a goal that I keep to myself). But you are the boss ;)

Comment (ddsol) [inline bin/index.js]: Chalk detects if color support is available, and perhaps more importantly, has multi-platform support / deals with special cases.

**Comment by ivanseidel** [comment]
Just did @ddsol. Let me know of anything else.

File: bin/index.js [hunk 4]
```diff
@@ -1,145 +1,164 @@
...
-const RESET = '\033[0m'
-const WHITE = '\033[37m'
-const YELLOW = '\033[33m'
-
-var step = 'Ping'
+var DraftLog = require('draftlog').into(console).addLineListener(process.stdin);
+var SpeedTestNet = require('../');
+
+/*
+ * Simple ANSI escape commands for Collors
```
Comment (ddsol) [inline bin/index.js]: Spelling `collors` => `colors`

File: bin/index.js [hunk 5]
```diff
@@ -1,145 +1,164 @@
...
   Ping: true,
   Download: false,
   Upload: false,
 }
 
-var header, speeds, location
-
-var halfwidth = 12
+/*
+ * Renders the header and Speeds (one beloow the other)
```
Comment (ddsol) [inline bin/index.js]: Spelling `beloow` => `below`

File: bin/index.js [hunk 6]
```diff
@@ -1,145 +1,164 @@
...
-  header('│' + headerTxt + '│')
-  speeds('│' + speedsTxt + '│')
+  header('│' + headerTxt + '│');
+  speeds('│' + speedsTxt + '│');
 }
 
-var frames = ['⠁', '⠂', '⠄', '⡀', '⢀', '⠠', '⠐', '⠈' ]
-var lastChange = 0
+/*
+ * Function that return state of Spinner, and change it's state with time
```
Comment (ddsol) [inline bin/index.js]: Spelling `return` => `returns`, `it's` => `its`

File: bin/index.js [hunk 7]
```diff
@@ -1,145 +1,164 @@
...
 test.on('uploadspeedprogress', function (speed){
-  if (!completedUpload)
-    statuses.Upload = speedText(speed, true)
+  if (!completedUpload) {
+    statuses.Upload = speedText(speed, true);
+  }
 })
+
 test.once('testserver', function (info){
   // Round to 1 decimal plate
```
Comment (ddsol) [inline bin/index.js]: Spelling `plate` => `place`

File: index.js [hunk 8]
```diff
@@ -817,52 +819,63 @@ function visualSpeedTest(options, callback) {
     log('Testing from ' + client.ip + ' at ' + client.isp + ', expected dl: ' + (client.ispdlavg / 8000).toFixed(2) + 'MB/s, expected ul: ' + (client.ispulavg / 8000).toFixed(2) + 'MB/s');
   });
 
-  function prog(what, pct) {
-    if (pct >= 100) {
-      if (bar) bar.terminate();
-      bar = null;
-      return;
-    }
+  var red   = '\u001b[41m \u001b[0m';
```
Comment (ddsol) [inline index.js]: Since we already have `chalk`, can we use it here as well?

**Comment by ivanseidel** [comment]
I wanted to test in a terminal that doesn't support colors nor any ANSI command. Do you have any idea of a good way to test that in iTerm or Terminal in MacOsx?

**Comment by ddsol** [comment]
No I don't. I'm on Windows and so it's not easy for me to test it... or to even know how to! I may log in to a server to test it from Linux however.

**Comment by ivanseidel** [comment]
My friend tried DraftLog in Windows normal Shell and it worked. Maybe it works for you

File: bin/index.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  var headerTxt = '';
+  var speedsTxt = '';
+  var spinner = Spinner();
+
+  // Build Header
+  for (var k in statuses) {
+    var status = statuses[k];
+
+    // Create header
+    var spacing = ' '.repeat(halfwidth - k.length / 2);
```
Comment (ddsol) [inline bin/index.js]: [Dual declare](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R60)

File: bin/index.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    if (!status) {
+      status = spinner + ' ';
+    }
+
+    // Create status text
+    var spacing = ' '.repeat(halfwidth - status.length / 2);
+
+    // If it's the current step, show as yellow
+    if (step == k) {
+      var txt = spacing + chalk.yellow(status) + spacing;
```
Comment (ddsol) [inline bin/index.js]: [Split declaration](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R66)

File: bin/index.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+ * Renders the header and Speeds (one below the other)
+ */
+var halfwidth = 12;
+function updateLines() {
+  var headerTxt = '';
+  var speedsTxt = '';
+  var spinner = Spinner();
+
+  // Build Header
+  for (var k in statuses) {
```
Comment (ddsol) [inline bin/index.js]: This variable is [declared twice](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R45)

File: bin/index.js [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    var status = statuses[k];
```
Comment (ddsol) [inline bin/index.js]: [Schizophrenic declaration](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R46)

File: bin/index.js [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+  speeds('│' + speedsTxt + '│');
+}
+
+/*
+ * Function that return state of Spinner, and change its state with time
+ */
+var frames = ['⠁', '⠂', '⠄', '⡀', '⢀', '⠠', '⠐', '⠈' ];
+var lastChange = 0;
+function Spinner(){
+  if (Date.now() > lastChange + 0) {
```
Comment (ddsol) [inline bin/index.js]: Could you remove the `+ 0`?

File: bin/index.js [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+console.log('│' + ' '.repeat(halfwidth * 6) + '│');
+console.log('└' + '─'.repeat(halfwidth * 6) + '┘');
+console.log();
+console.log();
+
+/*
+ * Start speed test
+ */
+var test = SpeedTestNet({maxTime: 10000});
+var interval = setInterval(updateLines, 60);
```
Comment (ddsol) [inline bin/index.js]: The `interval` variable isn't used anywhere so it's probably best if we get rid of it. Also, could you increase the interval to 100ms? It makes the spinner less frantic and the console more stable (less blinking).

File: bin/index.js [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+var completedUpload = false;
+var completedDownload = false;
+
+test.on('downloadspeedprogress', function (speed){
+  if (!completedDownload) {
+    statuses.Download = speedText(speed, true);
+  }
+})
```
Comment (ddsol) [inline bin/index.js]: semi

File: bin/index.js [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+test.on('uploadspeedprogress', function (speed){
+  if (!completedUpload) {
+    statuses.Upload = speedText(speed, true);
+  }
+})
```
Comment (ddsol) [inline bin/index.js]: semi

File: bin/index.js [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+  var title = info.sponsor + ', ' + info.country + ' - ' + info.name;
+  title = title + (title.length % 2 == 0 ? '' : ' ');
+
+  var fill = ' '.repeat(halfwidth * 3 - title.length / 2);
+  location('│' + fill + chalk.yellow(title) + fill + '│');
+  
+  var ping = Math.round(info.bestPing * 10) / 10;
+  statuses.Ping = ping + ' ms';
+  step = 'Download';
+})
```
Comment (ddsol) [inline bin/index.js]: Etc

File: index.js [hunk 18]
```diff
@@ -787,13 +786,18 @@ function speedTest(options) {
 module.exports = speedTest;
 
 function visualSpeedTest(options, callback) {
+  // We only need chalk and DraftLog here. Lazy load it.
+  var chalk = require('chalk')
```
Comment (ddsol) [inline index.js]: Semification

File: index.js [hunk 19]
```diff
@@ -787,13 +786,18 @@ function speedTest(options) {
...
+  var DraftLog = require('draftlog').into(console)
```
Comment (ddsol) [inline index.js]: Variable isn't used anywhere so please remove it. Also, a semi for consistency please.

File: index.js [hunk 20]
```diff
@@ -817,52 +821,68 @@ function visualSpeedTest(options, callback) {
...
-    if (pct >= 100) {
-      if (bar) bar.terminate();
-      bar = null;
-      return;
-    }
+  if (chalk.supportsColor) {
+    var red   = chalk.bgRed(' ');
+    var green = chalk.bgGreen(' ');
+  } else {
+    var red   = '─';
```
Comment (ddsol) [inline index.js]: This variable was already declared a few lines back.

File: .npmignore [hunk 21]
```diff
@@ -10,3 +10,4 @@ test
 .wafpickle-*
 CVS
 npm-debug.log
+media
```
Comment (ddsol) [inline .npmignore]: Please finish each file with a newline. Some tools have issues otherwise.

File: bin/index.js [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+  statuses.Download = speedText(speed);
+})
+
+test.once('uploadspeed', function (speed){
+  completedUpload = true;
+  step = 'Finished';
+  statuses.Upload = speedText(speed);
+})
+
+function speedText(speed, notFinished) {
```
Comment (ddsol) [inline bin/index.js]: `notFinished` isn't used. Can we remove it? (Also [here](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R119) and [here](https://github.com/ddsol/speedtest.net/pull/35/files#diff-9ea21889e9fbc7d82e7dabb0fb280808R125))

File: bin/index.js [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+})
+
+test.once('testserver', function (info){
+  // Round to 1 decimal place
+  var title = info.sponsor + ', ' + info.country + ' - ' + info.name;
+  title = title + (title.length % 2 == 0 ? '' : ' ');
+
+  var fill = ' '.repeat(halfwidth * 3 - title.length / 2);
+  location('│' + fill + chalk.yellow(title) + fill + '│');
+
```
Comment (ddsol) [inline bin/index.js]: Please remove spaces from this line

**Comment by ivanseidel** [comment]
Should be ok

File: bin/index.js [hunk 24]
```diff
@@ -132,7 +132,12 @@ function speedText(speed) {
 /*
  * Function that return state of Spinner, and change its state with time
  */
-var frames = ['⠁', '⠂', '⠄', '⡀', '⢀', '⠠', '⠐', '⠈' ];
+var frames = [
```
Comment (ddsol) [inline bin/index.js]: These fancy dots don't work in Windows. Could you add a Windows version?
```
var frames = process.platform === 'win32' ? [
    '|',
    '/',
    '-',
    '\\'
  ] : [
    '⠂⠁⠂⠄',
    '⠄⠂⠁⠂',
    '⠂⠄⠂⠁',
    '⠁⠂⠄⠂',
  ];
```
Or if you think of something fancier that doesn't require Unicode, then that'll work for me as well.

Comment (ivanseidel) [inline bin/index.js]: Check now

**Comment by ddsol** [comment]
The gif is now outdated 😕.

**Comment by ivanseidel** [comment]
Haha
I will do another one ;) give me a min.

**Comment by ivanseidel** [comment]
Done @ddsol

**Pull request merged**